src/nnet3/nnet-example-utils.cc

```diff
@@ -219,5 +219,200 @@ void GetComputationRequest(const Nnet &nnet,
+  const int32 extra_right_frames = num_input_frames_per_chunk - old_chunk_size -
+                  left_context - right_context - extra_left_frames;
+  KALDI_VLOG(2) << "extra_left_context=" << extra_left_frames 
+                << ", extra_right_context=" << extra_right_frames;
+  const int32 num_minibatches = split_egs->size();
+
+  for (int32 i = 0; i < num_minibatches; i++) {
+    std::vector<GeneralMatrix const*>  output_list(num_chunks);
+    std::vector<GeneralMatrix> output_matrices(num_chunks);
+
+    NnetExample &new_eg = (*split_egs)[i];
+    new_eg.io[feat].name = eg.io[feat].name;
+    // new chunks in the first/last minibatches should also include all extra
+    // frames before/after left_context/right_context from the old chunks
+    // (e.g. chunk_left_context/chunk_right_context)
+    new_eg.io[feat].indexes.resize(num_chunks * (new_chunk_size + left_context +
+              right_context + (i == 0 ? extra_left_frames : 0) +
+              (i == num_minibatches - 1 ? extra_right_frames : 0)));
+    for (int32 n = 0; n < num_chunks; n++) {
+      int32 src_begin = n * num_input_frames_per_chunk + i * new_chunk_size +
+                (i == 0 ? 0 : extra_left_frames),
+            src_end = src_begin + new_chunk_size + left_context + right_context +
+                (i == 0 ? extra_left_frames : 0) +
+                (i == num_minibatches - 1 ? extra_right_frames : 0),
+            dst_begin = n * (new_chunk_size + left_context + right_context +
+                (i == 0 ? extra_left_frames : 0) +
+                (i == num_minibatches - 1 ? extra_right_frames : 0));
+      // the last frame being copied should be the last row of features
+      if (i == num_minibatches - 1 && n == num_chunks - 1)
+        KALDI_ASSERT(src_end == eg.io[feat].features.NumRows());
+      //copy indexes
+      std::copy(eg.io[feat].indexes.begin() + src_begin,
+                eg.io[feat].indexes.begin() + src_end,
+                new_eg.io[feat].indexes.begin() + dst_begin);
+      // modify indexes "t"
+      int32 t = -left_context - (i == 0 ? extra_left_frames : 0);
+      for (int32 j = 0; j < src_end - src_begin; j++, t++)
+        new_eg.io[feat].indexes[dst_begin + j].t = t + output_t_begin[n];
+      // copy corresponding features
+      FilterGeneralMatrixRowsRange(eg.io[feat].features, src_begin,
+                                   src_end, &output_matrices[n]);
+      output_list[n] = &output_matrices[n];
+    }
+    AppendGeneralMatrixRows(output_list, &(new_eg.io[feat].features));
+  }
+}
+
+// Split ivector in eg and store them in *split_egs. It is called by
+// SplitChunk(). It assumes that the space of *split_egs has been allocated by
+// the caller.
+static void SplitIvector(int32 new_chunk_size, int32 num_chunks, int32 feat,
+                         const NnetExample &eg,
+                         std::vector<NnetExample> *split_egs) {
+  KALDI_ASSERT(NumFramesPerChunk(eg.io[feat]) == 1);
```

Comment:
Now that I think about it I think this might be the reason for the bad performance when doing state preserving training. You are using one i-vector for the entire 200 time steps. This is not advisable. Modify nnet3-get-egs to dump more ivectors per eg and use the ivector from the closest time step during the split.